crypto/property/property.c

```diff
@@ -0,0 +1,507 @@
+{
+    ALGORITHM *alg = NULL;
+    int i;
+
+    if (nid <= 0 || implementation == NULL)
+        return 0;
+    if (impls == NULL)
+        impls = g_implementations;
+
+    /* Only a read lock because no attempt is made to clean up empty stacks */
+    ossl_property_write_lock(impls);
+    ossl_impl_cache_flush(impls, nid);
+    alg = ossl_impl_store_retrieve(impls, nid);
+    if (alg == NULL) {
+        ossl_property_unlock(impls);
+        return 0;
+    }
+
+    /*
+     * A sorting find then a delete could be faster but these stacks should be
+     * relatively small, so we avoid the overhead.  Sorting could also surprise
+     * users when result orderings change (even though they are not guaranteed).
+     */
+    for (i = 0; i < sk_IMPLEMENTATION_num(alg->impls); i++) {
+        IMPLEMENTATION *impl = sk_IMPLEMENTATION_value(alg->impls, i);
+
+        if (impl->implementation == implementation) {
+            sk_IMPLEMENTATION_delete(alg->impls, i);
+            ossl_property_unlock(impls);
+            impl_free(impl);
+            return 1;
+        }
+    }
+    ossl_property_unlock(impls);
+    return 0;
+}
+
+int ossl_impl_store_fetch(OSSL_IMPL_STORE *impls,
+                          int nid, const char *properties,
+                          const void **result)
+{
+    ALGORITHM *alg;
+    IMPLEMENTATION *impl;
+    PROPERTY_LIST *pq = NULL, *p2;
+    int ret = 0;
+    int j;
+
+    if (nid <= 0 || result == NULL)
+        return 0;
+    if (impls == NULL)
+        impls = g_implementations;
+
+    /*
+     * This only needs to be a read lock, because queries never create property
+     * names or value and thus don't modify any of the property string layer.
+     */
+    ossl_property_read_lock(impls);
+    alg = ossl_impl_store_retrieve(impls, nid);
+    if (alg == NULL) {
+        ossl_property_unlock(impls);
+        return 0;
+    }
+
+    if (properties == NULL) {
+        if ((impl = sk_IMPLEMENTATION_value(alg->impls, 0)) != NULL) {
+            *result = impl->implementation;
+            ret = 1;
+        }
+        goto fin;
+    }
+    pq = ossl_parse_query(properties);
+    if (pq == NULL)
+        goto fin;
+    if (impls->global_properties != NULL) {
+        p2 = ossl_property_merge(pq, impls->global_properties);
+        if (p2 == NULL)
+            goto fin;
+        pq = p2;
+    }
+    for (j = 0; j < sk_IMPLEMENTATION_num(alg->impls); j++) {
+        impl = sk_IMPLEMENTATION_value(alg->impls, j);
+
+        if (ossl_property_compare(pq, impl->properties)) {
+            *result = impl->implementation;
+            ret = 1;
+            goto fin;
+        }
+    }
+fin:
+    ossl_property_unlock(impls);
+    ossl_property_free(pq);
+    return ret;
+}
+
+int ossl_impl_store_set_global_properties(OSSL_IMPL_STORE *impls,
+                                          const char *prop_query) {
+    int ret = 0;
+
+    if (impls == NULL)
+        return 1;
```

Comment:
Why not in `g_implementations`?

Reply:
It should be there, but as mentioned earlier, I suspect it is better if `g_implementations` disappears completely.

Reply:
I've removed `g_implementations` entirely.  The library context will need to provide.